pandas/io/formats/format.py

```diff
@@ -565,315 +569,414 @@ def __init__(
             )
-        self.na_rep = na_rep
-        self.decimal = decimal
+
+    def _initialize_justify(self, justify: Optional[str]) -> str:
+        if justify is None:
+            return get_option("display.colheader_justify")
+        else:
+            return justify
+
+    def _initialize_columns(self, columns: Optional[Sequence[str]]) -> Index:
+        if columns is not None:
+            cols = ensure_index(columns)
+            self.frame = self.frame[cols]
+            return cols
+        else:
+            return self.frame.columns
+
+    def _initialize_colspace(
+        self, col_space: Optional[ColspaceArgType]
+    ) -> ColspaceType:
+        result: ColspaceType
+
         if col_space is None:
-            self.col_space = {}
+            result = {}
         elif isinstance(col_space, (int, str)):
-            self.col_space = {"": col_space}
-            self.col_space.update({column: col_space for column in self.frame.columns})
+            result = {"": col_space}
+            result.update({column: col_space for column in self.frame.columns})
         elif isinstance(col_space, Mapping):
             for column in col_space.keys():
                 if column not in self.frame.columns and column != "":
                     raise ValueError(
                         f"Col_space is defined for an unknown column: {column}"
                     )
-            self.col_space = col_space
+            result = col_space
         else:
-            if len(frame.columns) != len(col_space):
+            if len(self.frame.columns) != len(col_space):
                 raise ValueError(
                     f"Col_space length({len(col_space)}) should match "
-                    f"DataFrame number of columns({len(frame.columns)})"
+                    f"DataFrame number of columns({len(self.frame.columns)})"
                 )
-            self.col_space = dict(zip(self.frame.columns, col_space))
+            result = dict(zip(self.frame.columns, col_space))
+        return result
 
-        self.header = header
-        self.index = index
-        self.line_width = line_width
-        self.max_rows = max_rows
-        self.min_rows = min_rows
-        self.max_cols = max_cols
-        self.max_rows_displayed = min(max_rows or len(self.frame), len(self.frame))
-        self.show_dimensions = show_dimensions
-        self.table_id = table_id
-        self.render_links = render_links
+    @property
+    def max_rows_displayed(self) -> int:
+        return min(self.max_rows or len(self.frame), len(self.frame))
 
-        if justify is None:
-            self.justify = get_option("display.colheader_justify")
-        else:
-            self.justify = justify
+    @property
+    def max_cols_fitted(self) -> Optional[int]:
+        """Number of columns fitting the screen."""
+        self._max_cols_fitted: Optional[int]
 
-        self.bold_rows = bold_rows
-        self.escape = escape
+        if hasattr(self, "_max_cols_fitted"):
```

Comment:
hasattr checks don't play play well with mypy

max_rows_displayed is called from is_truncated_horizontally which is quite prevalent.

optimisation here may be premature here, but in I think generally avoiding hasattr checks is preferable is called often.

```
class MyClass:
    pass

cls = MyClass()

%timeit hasattr(cls, "_max_cols_fitted")
# 120 ns ± 8.08 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

class MyClass:
    _max_cols_fitted = None

cls = MyClass()

%timeit cls._max_cols_fitted is not None
# 85.5 ns ± 1.94 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

what is the reason for changing `max_rows_adj` to a property?

Reply:
The reason for making the property out of ``max_rows_adj/max_rows_fitted`` was to place the computation inside the function. Before that it was scattered in the other methods, while it was difficult to understand how the calculation was actually carried out. The present implementation makes it more clear.

Before hasattr check I used try/except AttributeError check. But if I understood correctly, @jreback suggested to revert to hasattr check, which was there originally.

Reply:
having the separate function is good. maybe _calc_max_cols_fitted and just call it once and assign to attribute like before

> Before hasattr check I used try/except AttributeError check. But if I understood correctly, @jreback suggested to revert to hasattr check, which was there originally.

it was only called once before.

Reply:
If performance is a concern, we can consider ``cached_property`` decorator or it's equivalent created for Python 3.7. I can do this if required.

Reply:
> 
> 
> having the separate function is good. maybe _calc_max_cols_fitted and just call it once and assign to attribute like before
> 
> > Before hasattr check I used try/except AttributeError check. But if I understood correctly, @jreback suggested to revert to hasattr check, which was there originally.
> 
> it was only called once before.

This is fine with me. I will change as you suggest.

Reply:
> If performance is a concern

it's not. just suggesting avoiding certain patterns

from https://github.com/python/mypy/issues/1424#issuecomment-223795437

> In general i despise hasattr() checks; it's usually much better to add a
class level initialization to None and check for that.

Reply:
> we can consider `cached_property` decorator or it's equivalent created for Python 3.7. I can do this if required.

This is not in the standard library til 3.8?

Reply:
It is in the standard library starting from 3.8. Before that there wasn't one. So, I was thinking could b reasonable to implement the cached property for the needs, if the version < 3.8.